### PR TITLE
Optimize format_args placeholders without options: Display::simple_fmt

### PR DESCRIPTION
**Pull request by m-ou-se**
This is part of #99012

`format_args!("{}", "hello")` pulls in the entire `Display for str` implementation, which includes the code necessary to support formatting options like padding (e.g. `"{:50^}"`, etc.), which is quite unnecessary for the basic case of formatting with no options (`"{}"`).

This adds a new method to the format trait: `Display::simple_fmt`. It is implemented by forwarding to the regular `Display::fmt` method, but is overridden in the `Display` impl of `String` and `str` to just call `write_str` directly, avoiding pulling in any code related to padding.

### CODE REVIEW

**Comment by m-ou-se** [comment]
This basic test program shows great results with this change:

```rust
#![feature(rustc_private)]
#![feature(lang_items)]
#![feature(start)]
#![no_std]

extern crate libc;

use core::fmt::{self, Write};

struct Stdout;

impl Write for Stdout {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        unsafe { libc::write(1, s.as_ptr().cast(), s.len()) };
        Ok(())
    }
}

#[start]
fn main(_: isize, _: *const *const u8) -> isize {
    writeln!(Stdout, "Hello, {}!", "world").is_err() as isize
}

#[lang = "eh_personality"]
extern "C" fn eh_personality() {}

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    unsafe { libc::abort() };
}
```

**Before:** A .text section of **5995 bytes**

**After:** A .text section of **2666 bytes** (A 56% reduction! :sparkles:)

Most of the difference is that the binary no longer includes `core::fmt::Formatter::pad` (1064 bytes) and `core::str::count::do_count_chars` (1491 bytes).

**Comment by m-ou-se** [comment]
For binaries that use string formatting with options anywhere, there won't be any benefit in binary size. There might still be a tiny tiny performance gain, although that's also unlikely.

But for small (e.g. embedded) programs, this can make a great difference, as the test program above shows.

**Comment by Amanieu** [comment]
Wouldn't it be simpler to have a default method on `Display` instead of a separate trait?

**Comment by m-ou-se** [comment]
> Wouldn't it be simpler to have a default method on `Display` instead of a separate trait? This could then be extended to work with other traits such as `Debug` and `UpperHex`.

Yes, definitely. (Depending on your definition of "simpler".) Doing it as a default method on the traits doesn't require specialization and is a better solution in various ways, but does require a bit more work to implement, as it requires more changes to the format_args!() macro and ArgumentV1 type, so I didn't do that in my draft implementation.

Now that we've verified this PR can provide a big improvement in some cases, I'll update it to do it the better way. :)

**Comment by m-ou-se** [comment]
It's now implemented for all format traits (Display, Debug, Binary, Octal, and so on), but that does increase the size of the vtable for those traits. Let's wait for the test results to see if that has any significant impact.

**Comment by m-ou-se** [comment]
@aDotInTheVoid this PR is failing on `[rustdoc-json] src/test/rustdoc-json/traits/uses_extern_trait.rs`, which says "FIXME(adotinthevoid): Theses shouldn't be here": 

https://github.com/rust-lang/rust/blob/c372b1470109547e37dcc29ffd4723ac9f66f15d/src/test/rustdoc-json/traits/uses_extern_trait.rs#L4

What is that test for?

**Comment by aDotInTheVoid** [comment]
**TLDR**: The test is fragile, and assumes `Debug` only has one item, will send fix pr)

> What is that test for?

The test is checking that when rustdoc-json adds a foreign trait to the index, it also adds the methods of the trait. The fixme is because long term we don't want to add foreign traits to the index, and have the index only have local items.

The test is failing because it trys to get the `fmt` method from the `Debug` trait, and then check that their is a method called `fmt` with that name. But the way it gets the `fmt` method is by assuming it is the only one in the Debug trait only has one method. When it encounters two methods, it hits an assertion and fails.

**Comment by m-ou-se** [comment]
> **TLDR**: The test is fragile, and assumes `Debug` only has one item, #105063 fixes this, sorry.

Thanks!

**Comment by m-ou-se** [comment]
> Regressions

There's quite a few ~1% regressions in compilation time. They're not that significant, but it would be nice to fix them.

My theory is that it takes LLVM a somewhat significant amount of time to basically optimize the default `simple_fmt` implementations to: `simple_fmt = fmt`, that is, use the same function (pointer) for both.

It would be nice if we had a feature for this in the language, so we can literally write `= Self::fmt;` as the default implementation, to make it an alias from the start. But we currently don't have that.

Maybe someone from the compiler team has some good ideas here. :)

**Comment by Amanieu** [comment]
I think this could be done as a MIR optimization when one function forwards directly to another. In that case we could just emit an LLVM symbol alias instead of emitting an inline function.

**Comment by nnethercote** [comment]
The [binary size](https://perf.rust-lang.org/compare.html?start=c372b1470109547e37dcc29ffd4723ac9f66f15d&end=c85ca3b2e89f3b73daa36b9976c48d69adbdd0cf&stat=size%3Alinked_artifact) results show lots of 1-2% regressions, which is unfortunate for this PR which is all about reducing binary sizes :(

**Comment by m-ou-se** [comment]
Nominating for the compiler team to discuss [this comment](https://github.com/rust-lang/rust/pull/104525#issuecomment-1332004100) and the one after.

Specifically:

- Can we do that as a MIR optimization, or in some other way that doesn't require a new language feature?
- Should we wait for that before merging this?

**Comment by saethlin** [comment]
MIR optimizations have changed significantly since that perf run. Just citing my own PRs:
https://github.com/rust-lang/rust/pull/107449
https://github.com/rust-lang/rust/pull/109247
https://github.com/rust-lang/rust/pull/110705
https://github.com/rust-lang/rust/pull/106428

It's also quite interesting to me the many of the debug regressions are entirely attributable to `run_linker`.

If the problem is trivial function wrappers, it is possible that the MIR optimization improvements above have entirely fixed the problem. I think it would be worth doing a new perf run against all the changes that have landed since the last one.

**Comment by nnethercote** [comment]
Much better perf results now :)  Still some sub-1% regressions in binary size. Not a showstopper, but a brief investigation would be good to see if they can be easily avoided.

**Comment by m-ou-se** [comment]
It's a bit unnecessary that the vtable used by `&dyn Display` now contains both `fmt` and `simple_fmt`. If it contains `fmt` anyway, there's no need for `simple_fmt` to be included as well. Adding `where Self: Sized` is the usual way of excluding something from the vtable, but that won't work in this case.

Putting it in a separate trait (so, `SimpleDisplay::fmt` instead of `Display::simple_fmt`) could avoid that, but that adds other complexity and requires specialization, making it less flexible.

**Comment by m-ou-se** [comment]
r? compiler

**Comment by petrochenkov** [comment]
>Putting it in a separate trait (so, `SimpleDisplay::fmt` instead of `Display::simple_fmt`) could avoid that, but that adds other complexity and requires specialization, making it less flexible.

How much complexity we are talking about?
And what is the flexibility needed for?

The optimization slightly pessimizes the common case to make a difference in very specific cases like https://github.com/rust-lang/rust/pull/104525#issuecomment-1318399783.
If the complexity is not large then maybe it makes sense to not pessimize the common case.
@rustbot author

**Comment by Kobzol** [comment]
Are there any situations where `simple_fmt` does something else than just call `write_str` with a dynamically produced string (e.g. `String/&str`) or a `'static` string (e.g. `bool`, enums with only fieldless variants)?

I wonder if we could instead add `fn as_str(&self) -> Option<&str>` to `Display`, similar to what `Arguments` already exposes.

With `as_str`, we could better optimize preallocation of the result string when using `format!` or `to_string`, because we would know the sizes of placeholders that implement `as_str`. This would not be possible with `simple_fmt`, I think.

The lowering would then have to match on `as_str` for simple placeholders, and either call `write_str` if it returns `Some`, or call the normal formatting machinery if it returns `None`.

**Comment by bjorn3** [comment]
`.as_str()` isn't possible for most types. Only strings can implement it.

> The lowering would then have to match on as_str for simple placeholders, and either call write_str if it returns Some, or call the normal formatting machinery if it returns None.

That will make binary bloat even worse as the if condition can't be optimized away due to the formatting machinery using dynamic dispatch everywhere.

**Comment by Kobzol** [comment]
> .as_str() isn't possible for most types. Only strings can implement it.

Not just strings, but also e.g. bools or enums with only fieldless variants - because the returned lifetime of `Option<&str>` can also be `'static` (AFAIK). My guess is that a similar set of types that can implement `simple_fmt` can also implement `as_str` - that's why I would be interested to see situations where `simple_fmt` != `write_str/as_str`.

But even if this was only applicable to strings, I think that it is an important case to optimize. I have seen many situations where strings are formatted, and now that's sadly suboptimal. The flattening/inlining of `format_args` has helped, but still it's enough to have a string in a variable (or in a `const`), and then it has to go through the whole `fmt` machinery needlessly.

> That will make binary bloat even worse as the if condition can't be optimized away due to the formatting machinery using dynamic dispatch everywhere.

Yeah, optimizing the branch would be problematic with dynamic dispatch. But I guess that this could be solved with calling a function through virtual dispatch that would do the condition inside, so that the condition is only codegened once. It would mean double indirection, but that would also happen for types having the default implementation of `simple_fmt`.

Something like
```rust
fn simple_fmt(obj: &Display, fmt: &mut Formatter) {
   match obj.as_str() {
     Some(s) => fmt.write_str(s),
     None => obj.fmt(fmt)
   }
}
```

**Comment by bjorn3** [comment]
> But I guess that this could be solved with calling a function through virtual dispatch that would do the condition inside, so that the condition is only codegened once.

That still doesn't allow eliminating the big `Display::fmt` implementations entirely for embedded systems where there are no uses of non-simple formatting, right?

**Comment by Kobzol** [comment]
> That still doesn't allow eliminating the big Display::fmt implementations entirely for embedded systems where there are no uses of non-simple formatting, right?

Probably not, and that is also not the point of the proposed `as_str` method, my motivation was to better optimize the runtime performance of formatting (amongst other things by better preallocating strings when doing `format!/to_string`).

Regarding binary size, I don't have much experience with Rust embedded, so this is just an uninformed opinion, but I *think* that the code size wins (althought they are great!) here might not translate that well into real world programs. If having fmt in the binary or not having fmt in the binary is a fundamental difference for an embedded program, then it seems quite brittle to base this on an optimization that basically breaks once you use `{:?}` or format something else than a string anywhere in the program. But maybe I'm wrong and there are use-cases for only formatting strings with `{}` everywhere in embedded programs, I'm not sure :)

**Comment by bjorn3** [comment]
> amongst other things by better preallocating strings when doing format!/to_string

Maybe add a size_hint method for that?

**Comment by Kobzol** [comment]
Nice, that would also work, and could be even more general, e.g. integers could return their log10 number of digits.

**Comment by JohnCSimon** [comment]
@m-ou-se 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
